### PR TITLE
Support mask(pmap)

### PR DESCRIPTION
**Pull request by j-towns**
This pr adds basic support for mask(pmap(fun)).

The initial motivation for this was to use masking to solve the 'last batch problem' with pmap, but it turns out that this was already do-able on master, by doing the opposite composition, pmap(mask(fun)):
```python                                                               
in_shapes = ['i']  # 'i' is the inner batch size                                                    
out_shape = ''                                                           
                                                                             
@partial(pmap, axis_name='outer_batch')                                        
@partial(mask, in_shapes=in_shapes, out_shape=out_shape)                 
def fun(x):                                                              
  return lax.psum(jnp.sum(x), 'outer_batch')  # Sum out ragged dim first       
                                                                             
args = [jnp.array([[1, 2], [5, 10]])]                                    
logical_env = dict(i=jnp.array([2, 1]))  # Total batch size == 3                   
                                                                             
padded_out = fun(args, logical_env)
# 1 + 2 + 5 = 8 so
# padded_out == array([8, 8])              
```

### CODE REVIEW

File: jax/core.py [hunk 1]
```diff
@@ -1579,6 +1582,26 @@ def axis_frame(axis_name):
     else:
       raise NameError("unbound axis name: {}".format(axis_name))
 
+  @contextmanager
+  def extend_axis_env_poly(axis_name, size: int):
```
Comment (mattjj) [inline jax/core.py:1586]: Can we just use the same `trace_state.axis_env` and `extend_axis_env` function?

File: jax/interpreters/masking.py [hunk 2]
```diff
@@ -432,6 +452,46 @@ def todo(vals):
...
+        axis_size=padded_axis_size)
+    shapes = tuple(shape[1:] if mapped_invar else shape
+                   for shape, mapped_invar in zip(shapes, mapped_invars))
+    logical_env_vals = tuple(logical_env[k] for k in env_keys)
+    padded_env = (env_keys, padded_env_vals)
+    # TODO(j-towns): if poly_axis_size happens to be a constant then just use
+    # mask_subtrace here
+    f, shapes_out = mapped_mask_subtrace(f, self.master, params['axis_name'],
+                                         poly_axis_size, shapes, padded_env)
+    if 'donated_invars' in params:
```
Comment (mattjj) [inline jax/interpreters/masking.py:478]: We can punt this to later, but I'd rather factor this like `pe.call_param_updaters` / `ad.call_param_updaters`.

File: jax/interpreters/masking.py [hunk 3]
```diff
@@ -432,6 +452,46 @@ def todo(vals):
...
+    poly_axis_size, = {shape[0]
+                       for shape, mapped_invar in zip(shapes, mapped_invars)
+                       if mapped_invar}
+    padded_axis_size = eval_poly(poly_axis_size, padded_env)
+    env_keys, padded_env_vals = unzip2(sorted(padded_env.items()))
+    params = dict(
+        params,
+        mapped_invars=((False,) * len(logical_env) + params['mapped_invars']),
+        axis_size=padded_axis_size)
+    shapes = tuple(shape[1:] if mapped_invar else shape
```
Comment (mattjj) [inline jax/interpreters/masking.py:470]: I think it's best to use `core.mapped_aval` here, in part because it handles units and in part because using that abstraction will help us revise things later.

**Review by mattjj — COMMENTED** [review]
@j-towns brilliant work as always. You're the best for pushing this forward.

I left a couple comments but the only one that I'd really like to follow up on before merging is whether we can just use the existing `trace_state.axis_env` / `extend_axis_env` instead of adding `trace_state.axis_env_poly` / `extend_axis_end_poly`. Can you say a bit about that? Was it just to keep masking stuff from interfering with anything else, or was there another reason?

File: jax/core.py (hunk 1 above)
Comment (j-towns) [reply jax/core.py:1586]: I think it's a little bit tricky to use the same `axis_env`. There are two options if you're going to do that:

1. Allow an AxisEnvFrame to contain _both_ a padded (hard) axis size and a polymorphic one. So for a polymorphic axis you'd have something like
    ```python
    trace_state.axis_env = [('i', 9, b ** 2 + a + 1)]
    ```
2. Have two AxisEnvFrames for polymorphic axes, one containing the hard axis size and the other containing the polymophic one, i.e.
    ```python
    trace_state.axis_env = [('i', 9),
                            ('i', b ** 2 + a + 1)]
    ```
Option 2 is a bit ugly IMO, option 1 is do-able but the thing that makes it a little bit complicated is that the hard size and the polymorphic size are pushed onto the stack from different places (hard size is [in pxla.py](https://github.com/google/jax/blob/41f1eb412c2522be34795cd47148f3e39405088c/jax/interpreters/pxla.py#L699), polymorphic size is [in masking.py](https://github.com/google/jax/blob/c88e28c9a51fb1e2a3d12eb4f5e70164f6143a04/jax/interpreters/masking.py#L115)). I think the polymorphic size gets pushed first, so we'd have to have some logic to replace the tuple ('i', b ** 2 + a + 1) with ('i', 9, b ** 2 + a + 1) when the hard size gets pushed. I'm gonna have a go at implementing that just to see what it looks like, but it will be slightly more complicated than the current version.

Comment (j-towns) [reply jax/core.py]: Going to resolve this but if you can see a better solution lmk and we can implement it!

**Comment by j-towns** [comment]
Converting to draft for now because I'm drafting some deeper changes to masking core which may make this implementation more straightforward and enable shapecheck(pmap).

**Comment by hawkinsp** [comment]
Closing as stale.
